### PR TITLE
Add faster approximate poisson random generator

### PR DESCRIPTION
**Pull request by Gattocrucco**
See #39312 .

Things in this PR that I predict you might want to discuss:
- moving `ndtr` to `jax._src.lax.special` to avoid an import loop
- the `poisson(..., method='approximate')` sampler supports float output and has a derivative; the derivative is not as well-defined as for a continuous distribution, so the estimates based on it are biased
- the new unit tests add 6% to the running time of `random_lax_test.py`

I don't use the derivative myself, I added it because I started out assuming all distributions in jax have one, then realized later that's not the case; no discrete distribution defines a derivative.

### CODE REVIEW

File: jax/_src/random/core.py [hunk 1]
```diff
@@ -1996,11 +1997,247 @@ def _poisson(key, lam, shape, dtype) -> Array:
...
+
+def _poisson_cdf_inversion(u, lam):
+  """Invert the cdf truncated to the first `_POISSON_NUM_TERMS` values."""
+  # unrolled pmf recurrence instead of cumsum(exp(logpmf(arange))): keeps every
+  # op elementwise at the broadcast shape, so xla fuses the inversion into one
+  # kernel instead of materializing shape x _POISSON_NUM_TERMS intermediates
+  pmf = lax.exp(-lam)
+  cdf = pmf
+  cdfs = [cdf]
+  for k in range(1, _POISSON_NUM_TERMS):
```
Comment (jakevdp) [inline jax/_src/random/core.py]: This is 22 flattened iterations, which may impact compile time. Maybe use `fori_loop` instead?

File: jax/_src/random/core.py [hunk 2]
```diff
@@ -1996,11 +1997,247 @@ def _poisson(key, lam, shape, dtype) -> Array:
...
+    pmf = pmf * (lam / k)
+    cdf = cdf + pmf
+    cdfs.append(cdf)
+
+  # re-normalize because of truncation and float rounding
+  u *= cdf
```
Comment (jakevdp) [inline jax/_src/random/core.py]: I would write `u *= cdfs[-1]` here to be more explicit that this is the value from the last loop iteration (rather than the reader of the code having to remember how Python variable scope works in order to understand the logic)

File: tests/random_lax_test.py [hunk 3]
```diff
@@ -918,6 +919,125 @@ def testPoissonCornerCases(self):
     samples = random.poisson(key, lam, shape=(3,))
     self.assertArraysEqual(samples, jnp.array([-1, 0, -1]), check_dtypes=False)
 
+  @jtu.sample_product(
+    lam=[0.5, 3, 9, 11, 50, 500],
+    dtype=jtu.dtypes.supported([np.int16, np.int32, np.int64]),
+  )
+  def testPoissonApproximate(self, lam, dtype):
```
Comment (jakevdp) [inline tests/random_lax_test.py]: Could we cover this by adding `method=["exact", "approximate"]` to the existing poisson test?

File: tests/random_lax_test.py [hunk 4]
```diff
@@ -918,6 +919,125 @@ def testPoissonCornerCases(self):
...
+    # The approximate method should still follow the Poisson distribution; its
+    # total variation error < 1e-4 is undetectable at this sample size.
+    key = lambda: self.make_key(0)
+    rand = lambda key, lam: random.poisson(key, lam, (10000,), dtype,
+                                           method='approximate')
+    crand = jax.jit(rand)
+
+    for samples in [rand(key(), lam), crand(key(), lam)]:
```
Comment (jakevdp) [inline tests/random_lax_test.py]: Note soon-to-be-merged changes to this kind of boilerplate in #38972

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good! Some initial comments – there's a tradeoff between compile time and runtime when it comes to deciding between `for` loops and `fori_loop`/`scan` for some of the operations used below. A challenge is that that tradeoff lands in different places for different devices/accelerators, but I think here it would be better to use a fori_loop

**Comment by Gattocrucco** [comment]
I considered switching to `fori_loop` or vectorized expressions but kept the unrolled version.

The unrolled version is indeed slower to compile, but it's so much more performant that I think it's better overall. The point of this whole alternative implementation is avoiding loops because they make everything slow.

Prompted by your observation, I optimized a bit the unrolled loop, reducing the number of iterations to the bare minimum.

Measurements fori_loop vs. unrolled:
- Compilation: fori_loop is 1.5x faster to compile on gpu, while on cpu it makes no difference.
- Memory: fori_loop uses 2-5x more memory. In particular, the unrolled loop is perfectly optimal on gpu: it does not allocate anything that scales with the output size apart from the output buffer.
- Speed: fori_loop is 10-20x slower on gpu, 2x slower on cpu.

**Comment by Gattocrucco** [comment]
Other measurement: the pre-existing poisson sampler `poisson(..., method='exact')` is 1.3x slower at compiling on gpu than the current unrolled `'approximate'` implementation, so the new sampler is faster at compiling than the pre-existing one.

**Comment by jakevdp** [comment]
Yesterday I added a tool to print tables summarizing sampler bias for changes like this: https://github.com/jax-ml/jax/blob/main/jax/tools/evaluate_sampler_bias.py

It doesn't support discrete distributions, but maybe we can update it with a viable statistic for the discrete case? It would be nice to use something like this to see whether there is any quantile-dependent bias in the new approximate sampler.

File: tests/random_lax_test.py [hunk 5]
```diff
@@ -818,6 +822,108 @@ def testPoissonCornerCases(self):
     samples = random.poisson(key, lam, shape=(3,))
     self.assertArraysEqual(samples, jnp.array([-1, 0, -1]), check_dtypes=False)
 
+  def testPoissonApproximateZeros(self):
+    key = self.make_key(0)
+    samples = random.poisson(key, 0.0, shape=(20,), method='approximate')
+    self.assertArraysEqual(samples, jnp.zeros_like(samples))
+
+  def testPoissonApproximateShape(self):
```
Comment (jakevdp) [inline tests/random_lax_test.py]: This could be covered by adding `poisson_approximate` to `RandomShapeTest` above

Comment (jakevdp) [reply tests/random_lax_test.py]: Added in #39409 – once it's in we could validate this approximate sampler with e.g.
```
$ python jax/tools/evaluate_sampler_bias.py --sampler=poisson --lam=5.0 --method=approximate
```

**Comment by Gattocrucco** [comment]
Merged the shape test into `RandomShapeTest`, and ran the verification script, summary:

```
┌────────┬──────┬──────────────────────┬─────────────┬─────────┐
│  lam   │ seed │ Max abs err (counts) │ Max rel err │ Max |z| │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│   0.01 │  1   │          0           │     0%      │  0.25   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│    0.1 │  2   │          0           │     0%      │  1.13   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│      1 │  3   │          0           │     0%      │  1.02   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│      3 │  4   │          0           │     0%      │  1.57   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│      7 │  5   │          0           │     0%      │  1.73   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│     10 │  6   │          0           │     0%      │  2.48   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│     30 │  7   │          0           │     0%      │  1.34   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│    100 │  8   │          1           │    1.30%    │  2.29   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│    300 │  9   │          0           │     0%      │  1.36   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│   1000 │  10  │          0           │     0%      │  1.51   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│  10000 │  11  │          1           │   0.010%    │  1.23   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│ 100000 │  12  │          1           │   0.0010%   │  1.90   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│   10^6 │  13  │          3           │   0.0003%   │  1.35   │
├────────┼──────┼──────────────────────┼─────────────┼─────────┤
│   10^7 │  14  │          11          │   0.0001%   │  1.41   │
└────────┴──────┴──────────────────────┴─────────────┴─────────┘
```

**Review by jakevdp — APPROVED** [review]
Thanks!

File: tests/random_lax_test.py [hunk 6]
```diff
@@ -826,6 +832,97 @@ def testPoissonCornerCases(self):
...
+  def testPoissonApproximateRngImpl(self):
+    # Unlike 'exact', the 'approximate' method only draws normal variates, so
+    # it works with any RNG implementation.
+    key = jax.random.key(0, impl='rbg')
+    with self.assertRaises(NotImplementedError):
+      random.poisson(key, 3.0, (100,))
+    samples = random.poisson(key, 3.0, (100,), method='approximate')
+    self.assertTrue(jnp.all(samples >= 0))
+
+  @jtu.sample_product(lam=[0.5, 2., 7., 20., 1e3])
```
Comment (jakevdp) [inline tests/random_lax_test.py]: I think we could probably remove the tests from here down. I don't think we should promise anything about the gradients of a discrete sampler.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: tests/random_lax_test.py (hunk 6 above)
Comment (Gattocrucco) [reply tests/random_lax_test.py]: Questions:
- There are 2 tests on gradients and 1 on approximation accuracy down there, do you mean to remove the one on accuracy as well?
- I think either we keep the `custom_jvp` and test it, or drop it. No derivation without verification.

Derivatives of a discrete sampler are somewhat a handle-with-care thing, so if I was you, and I was not paying particular attention to what's entering the codebase with this PR, I'd want the derivatives removed. Just saying.

Comment (jakevdp) [reply tests/random_lax_test.py]: I think dropping the custom JVP along with the gradient-related tests is the best option. Gradients for a discrete sampler don't strike me as particularly important.

**Review by jakevdp — APPROVED** [review]
Thanks!
